### HW1 Создание и нормализация базы данных

### Задачи

* Продумать структуру БД и отрисовать в редакторе; (вып.)
* Нормализовать БД (1НФ-3НФ), описав, к какой нормальной форме приводится таблица и почему таблица в этой нормальной форме изначально не находилась; (вып.)
* Создать все таблицы в DBeaver, указав первичные ключи к таблицам, правильные типы данных, могут ли поля быть пустыми или нет;
* Загрузить данные в таблицы в соответствии с созданной структурой.

## Решение

### Проанализируем исходный датасет

Исходный датасет представлени в виде документа в формате .xlsx, в котором содержится две таблицы: 'transaction' и 'customer', следующего вида:

transaction:

![transaction](https://thumb.cloud.mail.ru/weblink/thumb/xw1/65YG/r4D73Erwp)

customer:

![customer](https://thumb.cloud.mail.ru/weblink/thumb/xw1/WFgw/USjtiCqFX)

Глядя на талицы можно выделить четыре основные сущности: transaction, customer, product и address (адреса лучше выделить в отдельный справочник так как у одного покупателя может быть несколько адресов доставки)

Разделим таблицы по сущностям с помощью pandas, но сначала выделим признаки для каждой сущности:

* transaction: 

![transaction](https://thumb.cloud.mail.ru/weblink/thumb/xw1/KhME/TgaiUXQPE)

* customer:

![customer](https://thumb.cloud.mail.ru/weblink/thumb/xw1/xHRs/hdKoDUXXh)

* product:

![product](https://thumb.cloud.mail.ru/weblink/thumb/xw1/AhFT/ss4GmQHTp)


* address:

![address](https://thumb.cloud.mail.ru/weblink/thumb/xw1/nKBJ/hVQnbsPqk)


Колонки list_price и standard_cost оставляем в transaction и в product так как нам необходимо в транзакциях видеть цены на момент совершения покупки, а в product мы будем видеть актуальные цены на продукт

In [274]:
import pandas as pd
import openpyxl

Загрузим таблицу transaction:

In [275]:
transactions_df = pd.read_excel('HW1_data.xlsx')
transactions_df.head()

transaction_id  product_id  customer_id transaction_date online_order  \
0               1           2         2950       2017-02-25        False   
1               2           3         3120       2017-05-21         True   
2               3          37          402       2017-10-16        False   
3               4          88         3135       2017-08-31        False   
4               5          78          787       2017-10-01         True   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  
0       71.49          53.62  
1     2091.47         388.92  
2     1793.43         248.82  
3     1198.46         381.10  
4     1765.30         709.48

In [276]:
transactions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transaction_id    20000 non-null  int64         
 1   product_id        20000 non-null  int64         
 2   customer_id       20000 non-null  int64         
 3   transaction_date  20000 non-null  datetime64[ns]
 4   online_order      19640 non-null  object        
 5   order_status      20000 non-null  object        
 6   brand             19803 non-null  object        
 7   product_line      19803 non-null  object        
 8   product_class     19803 non-null  object        
 9   product_size      19803 non-null  object        
 10  list_price        20000 non-null  float64       
 11  standard_cost     19803 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(3), object(6)
memory usage: 1.8+ MB


Загрузим таблицу customer:

In [277]:
customers_df = pd.read_excel('HW1_data.xlsx', 1)
customers_df.head()

customer_id      first_name  last_name  gender                  DOB  \
0            1         Laraine  Medendorp       F  1953-10-12 00:00:00   
1            2             Eli    Bockman    Male  1980-12-16 00:00:00   
2            3           Arlin     Dearle    Male  1954-01-20 00:00:00   
3            4          Talbot        NaN    Male  1961-10-03 00:00:00   
4            5  Sheila-kathryn     Calton  Female  1977-05-13 00:00:00   

                job_title job_industry_category     wealth_segment  \
0     Executive Secretary                Health      Mass Customer   
1  Administrative Officer    Financial Services      Mass Customer   
2      Recruiting Manager              Property      Mass Customer   
3                     NaN                    IT      Mass Customer   
4           Senior Editor                   NaN  Affluent Customer   

  deceased_indicator owns_car              address  postcode            state  \
0                  N      Yes   060 Morning Avenue      2016  New South Wales   
1                  N      Yes  6 Meadow Vale Court      2153  New South Wales   
2                  N      Yes   0 Holy Cross Court      4211              QLD   
3                  N       No  17979 Del Mar Point      2448  New South Wales   
4                  N      Yes     9 Oakridge Court      3216              VIC   

     country  property_valuation  
0  Australia                  10  
1  Australia                  10  
2  Australia                   9  
3  Australia                   4  
4  Australia                   9

In [278]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   customer_id            4000 non-null   int64 
 1   first_name             4000 non-null   object
 2   last_name              3875 non-null   object
 3   gender                 4000 non-null   object
 4   DOB                    3913 non-null   object
 5   job_title              3494 non-null   object
 6   job_industry_category  3344 non-null   object
 7   wealth_segment         4000 non-null   object
 8   deceased_indicator     4000 non-null   object
 9   owns_car               4000 non-null   object
 10  address                4000 non-null   object
 11  postcode               4000 non-null   int64 
 12  state                  4000 non-null   object
 13  country                4000 non-null   object
 14  property_valuation     4000 non-null   int64 
dtypes: int64(3), object(1

### Разделим две исходные таблицы на 4 датасета по сущностям, которые мы выделили раньше:

In [279]:
transaction_df = transactions_df[['transaction_id', 'product_id', 'customer_id', 'transaction_date', 'online_order', 'order_status', 'list_price', 'standard_cost']]
transaction_df.head()

transaction_id  product_id  customer_id transaction_date online_order  \
0               1           2         2950       2017-02-25        False   
1               2           3         3120       2017-05-21         True   
2               3          37          402       2017-10-16        False   
3               4          88         3135       2017-08-31        False   
4               5          78          787       2017-10-01         True   

  order_status  list_price  standard_cost  
0     Approved       71.49          53.62  
1     Approved     2091.47         388.92  
2     Approved     1793.43         248.82  
3     Approved     1198.46         381.10  
4     Approved     1765.30         709.48

In [280]:
customer_df = customers_df[['customer_id', 'first_name', 'last_name', 'gender', 'DOB', 'job_title', 'job_industry_category', 'wealth_segment', 'deceased_indicator', 'owns_car']]
customer_df.head()

customer_id      first_name  last_name  gender                  DOB  \
0            1         Laraine  Medendorp       F  1953-10-12 00:00:00   
1            2             Eli    Bockman    Male  1980-12-16 00:00:00   
2            3           Arlin     Dearle    Male  1954-01-20 00:00:00   
3            4          Talbot        NaN    Male  1961-10-03 00:00:00   
4            5  Sheila-kathryn     Calton  Female  1977-05-13 00:00:00   

                job_title job_industry_category     wealth_segment  \
0     Executive Secretary                Health      Mass Customer   
1  Administrative Officer    Financial Services      Mass Customer   
2      Recruiting Manager              Property      Mass Customer   
3                     NaN                    IT      Mass Customer   
4           Senior Editor                   NaN  Affluent Customer   

  deceased_indicator owns_car  
0                  N      Yes  
1                  N      Yes  
2                  N      Yes  
3                  N       No  
4                  N      Yes

In [281]:
product_df = transactions_df[['product_id', 'brand', 'product_line', 'product_class', 'product_size', 'list_price', 'standard_cost']]
product_df.head()

product_id           brand product_line product_class product_size  \
0           2           Solex     Standard        medium       medium   
1           3   Trek Bicycles     Standard        medium        large   
2          37      OHM Cycles     Standard           low       medium   
3          88  Norco Bicycles     Standard        medium       medium   
4          78  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  
0       71.49          53.62  
1     2091.47         388.92  
2     1793.43         248.82  
3     1198.46         381.10  
4     1765.30         709.48

In [282]:
address_df = customers_df[['address', 'postcode', 'state', 'country', 'property_valuation']]
address_df.head()

address  postcode            state    country  \
0   060 Morning Avenue      2016  New South Wales  Australia   
1  6 Meadow Vale Court      2153  New South Wales  Australia   
2   0 Holy Cross Court      4211              QLD  Australia   
3  17979 Del Mar Point      2448  New South Wales  Australia   
4     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

### Разбираемся с ключами для каждой таблицы

## Transaction

Проверим transaction_id на пропуски и дубликаты:

In [283]:
transaction_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transaction_id    20000 non-null  int64         
 1   product_id        20000 non-null  int64         
 2   customer_id       20000 non-null  int64         
 3   transaction_date  20000 non-null  datetime64[ns]
 4   online_order      19640 non-null  object        
 5   order_status      20000 non-null  object        
 6   list_price        20000 non-null  float64       
 7   standard_cost     19803 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(3), object(2)
memory usage: 1.2+ MB


In [284]:
duplicate_transaction = transaction_df.duplicated(['transaction_id'])
duplicate_transaction.value_counts()

False    20000
Name: count, dtype: int64

Так как transaction_id не имеет дубликатов и пропусков принимаем данный признак за первичный ключ для таблицы transaction

## Customer

Проверим customer_id на пропуски и дубликаты:

In [285]:
customer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   customer_id            4000 non-null   int64 
 1   first_name             4000 non-null   object
 2   last_name              3875 non-null   object
 3   gender                 4000 non-null   object
 4   DOB                    3913 non-null   object
 5   job_title              3494 non-null   object
 6   job_industry_category  3344 non-null   object
 7   wealth_segment         4000 non-null   object
 8   deceased_indicator     4000 non-null   object
 9   owns_car               4000 non-null   object
dtypes: int64(1), object(9)
memory usage: 312.6+ KB


In [286]:
duplicate_customer = customer_df.duplicated(['customer_id'])
duplicate_customer.value_counts()

False    4000
Name: count, dtype: int64

Так как customer_id не имеет дубликатов и пропусков принимаем данный признак за первичный ключ для таблицы customer

## Product

Проверим product_id на пропуски и дубликаты:

In [287]:
product_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     20000 non-null  int64  
 1   brand          19803 non-null  object 
 2   product_line   19803 non-null  object 
 3   product_class  19803 non-null  object 
 4   product_size   19803 non-null  object 
 5   list_price     20000 non-null  float64
 6   standard_cost  19803 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 1.1+ MB


In [288]:
duplicate_product = product_df.duplicated(['product_id'])
duplicate_product.value_counts()

True     19899
False      101
Name: count, dtype: int64

Так как product_id имеет много дубликатов мы не можем взять его как первичный ключ для таблицы product.
Попробуем удалить дубликаты в product_df

In [289]:
product_df_copy = product_df.drop_duplicates()

In [290]:
duplicate_product_copy = product_df_copy.duplicated(['product_id'])
duplicate_product_copy.value_counts()

True     298
False    101
Name: count, dtype: int64

In [291]:
product_df_copy.sample(10)

product_id           brand product_line product_class product_size  \
13519           0             NaN          NaN           NaN          NaN   
18244           0             NaN          NaN           NaN          NaN   
11548           0             NaN          NaN           NaN          NaN   
158            10        WeareA2B      Touring        medium       medium   
4024            0             NaN          NaN           NaN          NaN   
434             4           Solex     Standard        medium       medium   
889           100   Trek Bicycles     Standard        medium        small   
708            41  Norco Bicycles     Standard           low       medium   
13             16  Norco Bicycles     Standard          high        small   
6932            0             NaN          NaN           NaN          NaN   

       list_price  standard_cost  
13519      405.94            NaN  
18244     1780.22            NaN  
11548       74.89            NaN  
158       1466.68         363.25  
4024      1880.40            NaN  
434       1483.20          99.59  
889       1386.84        1234.29  
708        958.74         748.90  
13        1661.92        1479.11  
6932       172.09            NaN

In [292]:
product_df_copy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 399 entries, 0 to 19871
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     399 non-null    int64  
 1   brand          203 non-null    object 
 2   product_line   203 non-null    object 
 3   product_class  203 non-null    object 
 4   product_size   203 non-null    object 
 5   list_price     399 non-null    float64
 6   standard_cost  203 non-null    float64
dtypes: float64(2), int64(1), object(4)
memory usage: 24.9+ KB


Удалив дубликаты в датасете мы видим что для product_id	== 0 осталось очень много аномальных записей (отсутсвуют 5 из 6 признаков), данный датасет требует более детальной проработки так как аномальные записи не несут информативноссти для БД (но это не входит в данное ДЗ). Просто удалим строки где отсутствуют 5 из 6 признаков.

In [293]:
product_df_copy = product_df_copy[(product_df_copy['brand'].notna()) & (product_df_copy['product_line'].notna()) & (product_df_copy['product_class'].notna()) & (product_df_copy['product_size'].notna()) & (product_df_copy['standard_cost'].notna())]

Так же необходимо удалить дубликаты по комбинациям признаков без учёта цен на продукт, так как цены могут часто меняться, а в справочнике для продуктов необходимо иметь только уникальные комбинации для идентификации продукта.

In [294]:
mask = product_df_copy.duplicated(subset=['product_id', 'brand', 'product_line', 'product_class', 'product_size'])
product_df = product_df_copy.loc[~mask]

In [295]:
product_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 189 entries, 0 to 17468
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     189 non-null    int64  
 1   brand          189 non-null    object 
 2   product_line   189 non-null    object 
 3   product_class  189 non-null    object 
 4   product_size   189 non-null    object 
 5   list_price     189 non-null    float64
 6   standard_cost  189 non-null    float64
dtypes: float64(2), int64(1), object(4)
memory usage: 11.8+ KB


Итого в таблице product 189 строк с уникальными значениями

На основании анализа принимаем решение добавить для product_df суррогатный первичный ключ:

In [296]:
product_df = product_df.sort_values('product_id').reset_index(drop=True)

In [297]:
product_df = product_df.reset_index().rename(columns={'index': 'id'})
product_df.head()

id  product_id           brand product_line product_class product_size  \
0   0           0      OHM Cycles         Road        medium       medium   
1   1           0  Norco Bicycles     Standard           low       medium   
2   2           0      OHM Cycles         Road          high        large   
3   3           0   Trek Bicycles     Standard        medium       medium   
4   4           0      OHM Cycles     Standard        medium       medium   

   list_price  standard_cost  
0      742.54     667.400024  
1      363.01     290.410000  
2       12.01       7.210000  
3      499.53     388.720000  
4      235.63     125.070000

In [298]:
product_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 189 entries, 0 to 188
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             189 non-null    int64  
 1   product_id     189 non-null    int64  
 2   brand          189 non-null    object 
 3   product_line   189 non-null    object 
 4   product_class  189 non-null    object 
 5   product_size   189 non-null    object 
 6   list_price     189 non-null    float64
 7   standard_cost  189 non-null    float64
dtypes: float64(2), int64(2), object(4)
memory usage: 11.9+ KB


Теперь необходимо отредактировать внешний ключ product_id в основной таблице transactions

In [299]:
transaction_df = transactions_df.merge(product_df, on=['product_id', 'brand', 'product_line', 'product_class', 'product_size', 'list_price', 'standard_cost'], how='left')

In [300]:
transaction_df = transaction_df.drop(columns=['product_id', 'brand', 'product_line', 'product_class', 'product_size']).rename(columns={'id': 'product_id'})

In [301]:
transaction_df

transaction_id  customer_id transaction_date online_order order_status  \
0                   1         2950       2017-02-25        False     Approved   
1                   2         3120       2017-05-21         True     Approved   
2                   3          402       2017-10-16        False     Approved   
3                   4         3135       2017-08-31        False     Approved   
4                   5          787       2017-10-01         True     Approved   
...               ...          ...              ...          ...          ...   
19995           19996         1018       2017-06-24         True     Approved   
19996           19997          127       2017-11-09         True     Approved   
19997           19998         2284       2017-04-14         True     Approved   
19998           19999         2764       2017-07-03        False     Approved   
19999           20000         1144       2017-09-22         True     Approved   

       list_price  standard_cost  product_id  
0           71.49          53.62        21.0  
1         2091.47         388.92        22.0  
2         1793.43         248.82        82.0  
3         1198.46         381.10       164.0  
4         1765.30         709.48       147.0  
...           ...            ...         ...  
19995     2005.66        1203.40       104.0  
19996      416.98         312.74        89.0  
19997     1636.90          44.71       162.0  
19998      227.88         136.73        27.0  
19999     1775.81        1580.47        36.0  

[20000 rows x 8 columns]

## Address

В таблице adress_df изначально нет признака, который мы можем взять за первичный ключ. Поэтому разу сделаем суррогатный ключ.

In [302]:
address_df = address_df.reset_index().rename(columns={'index': 'address_id'})
address_df.head()

address_id              address  postcode            state    country  \
0           0   060 Morning Avenue      2016  New South Wales  Australia   
1           1  6 Meadow Vale Court      2153  New South Wales  Australia   
2           2   0 Holy Cross Court      4211              QLD  Australia   
3           3  17979 Del Mar Point      2448  New South Wales  Australia   
4           4     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [303]:
address_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   address_id          4000 non-null   int64 
 1   address             4000 non-null   object
 2   postcode            4000 non-null   int64 
 3   state               4000 non-null   object
 4   country             4000 non-null   object
 5   property_valuation  4000 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [304]:
duplicate_address = address_df.duplicated(['address_id'])
duplicate_address.value_counts()

False    4000
Name: count, dtype: int64

В колонке address указан полный адрес (дом и улица). Имеет смысл сразу разделить данный признак на два, так как в будущем это нужно будет сделать для нормализации БД.

In [305]:
import warnings
warnings.filterwarnings('ignore')

In [306]:
address_df['address'] = address_df['address'].map(lambda x: x.split(' ', 1))

In [307]:
address_df['house_number'] = address_df['address'].map(lambda x: x[0])

In [308]:
address_df['street'] = address_df['address'].map(lambda x: x[1])

In [309]:
address_df = address_df.drop('address', axis=1)

In [310]:
address_df

address_id  postcode            state    country  property_valuation  \
0              0      2016  New South Wales  Australia                  10   
1              1      2153  New South Wales  Australia                  10   
2              2      4211              QLD  Australia                   9   
3              3      2448  New South Wales  Australia                   4   
4              4      3216              VIC  Australia                   9   
...          ...       ...              ...        ...                 ...   
3995        3995      4511              QLD  Australia                   6   
3996        3996      2756              NSW  Australia                  10   
3997        3997      4032              QLD  Australia                   7   
3998        3998      2251              NSW  Australia                   7   
3999        3999      3996              NSW  Australia                  10   

     house_number               street  
0             060       Morning Avenue  
1               6    Meadow Vale Court  
2               0     Holy Cross Court  
3           17979        Del Mar Point  
4               9       Oakridge Court  
...           ...                  ...  
3995        57042  Village Green Point  
3996           87  Crescent Oaks Alley  
3997         8194          Lien Street  
3998          320          Acker Drive  
3999          613            Erie Lane  

[4000 rows x 7 columns]

Осталось добавить врешний ключ address_id в основную таблицу transaction

In [311]:
transaction_df

transaction_id  customer_id transaction_date online_order order_status  \
0                   1         2950       2017-02-25        False     Approved   
1                   2         3120       2017-05-21         True     Approved   
2                   3          402       2017-10-16        False     Approved   
3                   4         3135       2017-08-31        False     Approved   
4                   5          787       2017-10-01         True     Approved   
...               ...          ...              ...          ...          ...   
19995           19996         1018       2017-06-24         True     Approved   
19996           19997          127       2017-11-09         True     Approved   
19997           19998         2284       2017-04-14         True     Approved   
19998           19999         2764       2017-07-03        False     Approved   
19999           20000         1144       2017-09-22         True     Approved   

       list_price  standard_cost  product_id  
0           71.49          53.62        21.0  
1         2091.47         388.92        22.0  
2         1793.43         248.82        82.0  
3         1198.46         381.10       164.0  
4         1765.30         709.48       147.0  
...           ...            ...         ...  
19995     2005.66        1203.40       104.0  
19996      416.98         312.74        89.0  
19997     1636.90          44.71       162.0  
19998      227.88         136.73        27.0  
19999     1775.81        1580.47        36.0  

[20000 rows x 8 columns]

In [312]:
address_df

address_id  postcode            state    country  property_valuation  \
0              0      2016  New South Wales  Australia                  10   
1              1      2153  New South Wales  Australia                  10   
2              2      4211              QLD  Australia                   9   
3              3      2448  New South Wales  Australia                   4   
4              4      3216              VIC  Australia                   9   
...          ...       ...              ...        ...                 ...   
3995        3995      4511              QLD  Australia                   6   
3996        3996      2756              NSW  Australia                  10   
3997        3997      4032              QLD  Australia                   7   
3998        3998      2251              NSW  Australia                   7   
3999        3999      3996              NSW  Australia                  10   

     house_number               street  
0             060       Morning Avenue  
1               6    Meadow Vale Court  
2               0     Holy Cross Court  
3           17979        Del Mar Point  
4               9       Oakridge Court  
...           ...                  ...  
3995        57042  Village Green Point  
3996           87  Crescent Oaks Alley  
3997         8194          Lien Street  
3998          320          Acker Drive  
3999          613            Erie Lane  

[4000 rows x 7 columns]

In [313]:
transaction_df = transaction_df.merge(address_df, left_on='customer_id', right_on='address_id', how='left')

In [314]:
transaction_df = transaction_df.drop(columns=['postcode', 'state', 'country', 'property_valuation', 'house_number', 'street'])

In [315]:
transaction_df

transaction_id  customer_id transaction_date online_order order_status  \
0                   1         2950       2017-02-25        False     Approved   
1                   2         3120       2017-05-21         True     Approved   
2                   3          402       2017-10-16        False     Approved   
3                   4         3135       2017-08-31        False     Approved   
4                   5          787       2017-10-01         True     Approved   
...               ...          ...              ...          ...          ...   
19995           19996         1018       2017-06-24         True     Approved   
19996           19997          127       2017-11-09         True     Approved   
19997           19998         2284       2017-04-14         True     Approved   
19998           19999         2764       2017-07-03        False     Approved   
19999           20000         1144       2017-09-22         True     Approved   

       list_price  standard_cost  product_id  address_id  
0           71.49          53.62        21.0      2950.0  
1         2091.47         388.92        22.0      3120.0  
2         1793.43         248.82        82.0       402.0  
3         1198.46         381.10       164.0      3135.0  
4         1765.30         709.48       147.0       787.0  
...           ...            ...         ...         ...  
19995     2005.66        1203.40       104.0      1018.0  
19996      416.98         312.74        89.0       127.0  
19997     1636.90          44.71       162.0      2284.0  
19998      227.88         136.73        27.0      2764.0  
19999     1775.81        1580.47        36.0      1144.0  

[20000 rows x 9 columns]

Готово! Этого хватит чтобы выполнить первое задание, проектирование БД в редакторе.

### Проектируем БД в редакторе

Для проектирования используем dbdiagram.io

![dbdiagram](https://thumb.cloud.mail.ru/weblink/thumb/xw1/PPBP/vyHy1cRVU)

Скрипт:

![skript1](https://thumb.cloud.mail.ru/weblink/thumb/xw1/hbtN/pWCskGxFr)
![skript2](https://thumb.cloud.mail.ru/weblink/thumb/xw1/Dngh/Rv6ForN24)

Скрипт в папке data

# Нормализация

Берем 4 талицы, которые мы получили на предыдущем этапе, анализируя датасет

In [321]:
transaction_df

transaction_id  customer_id transaction_date online_order order_status  \
0                   1         2950       2017-02-25        False     Approved   
1                   2         3120       2017-05-21         True     Approved   
2                   3          402       2017-10-16        False     Approved   
3                   4         3135       2017-08-31        False     Approved   
4                   5          787       2017-10-01         True     Approved   
...               ...          ...              ...          ...          ...   
19995           19996         1018       2017-06-24         True     Approved   
19996           19997          127       2017-11-09         True     Approved   
19997           19998         2284       2017-04-14         True     Approved   
19998           19999         2764       2017-07-03        False     Approved   
19999           20000         1144       2017-09-22         True     Approved   

       list_price  standard_cost  product_id  address_id  
0           71.49          53.62        21.0      2950.0  
1         2091.47         388.92        22.0      3120.0  
2         1793.43         248.82        82.0       402.0  
3         1198.46         381.10       164.0      3135.0  
4         1765.30         709.48       147.0       787.0  
...           ...            ...         ...         ...  
19995     2005.66        1203.40       104.0      1018.0  
19996      416.98         312.74        89.0       127.0  
19997     1636.90          44.71       162.0      2284.0  
19998      227.88         136.73        27.0      2764.0  
19999     1775.81        1580.47        36.0      1144.0  

[20000 rows x 9 columns]

In [322]:
customer_df

customer_id      first_name  last_name  gender                  DOB  \
0               1         Laraine  Medendorp       F  1953-10-12 00:00:00   
1               2             Eli    Bockman    Male  1980-12-16 00:00:00   
2               3           Arlin     Dearle    Male  1954-01-20 00:00:00   
3               4          Talbot        NaN    Male  1961-10-03 00:00:00   
4               5  Sheila-kathryn     Calton  Female  1977-05-13 00:00:00   
...           ...             ...        ...     ...                  ...   
3995         3996         Rosalia   Halgarth  Female  1975-08-09 00:00:00   
3996         3997          Blanch     Nisuis  Female  2001-07-13 00:00:00   
3997         3998          Sarene    Woolley       U                  NaN   
3998         3999       Patrizius        NaN    Male  1973-10-24 00:00:00   
3999         4000           Kippy    Oldland    Male  1991-11-05 00:00:00   

                   job_title job_industry_category     wealth_segment  \
0        Executive Secretary                Health      Mass Customer   
1     Administrative Officer    Financial Services      Mass Customer   
2         Recruiting Manager              Property      Mass Customer   
3                        NaN                    IT      Mass Customer   
4              Senior Editor                   NaN  Affluent Customer   
...                      ...                   ...                ...   
3995   VP Product Management                Health      Mass Customer   
3996         Statistician II         Manufacturing     High Net Worth   
3997       Assistant Manager                    IT     High Net Worth   
3998                     NaN         Manufacturing  Affluent Customer   
3999    Software Engineer IV                   NaN  Affluent Customer   

     deceased_indicator owns_car  
0                     N      Yes  
1                     N      Yes  
2                     N      Yes  
3                     N       No  
4                     N      Yes  
...                 ...      ...  
3995                  N       No  
3996                  N      Yes  
3997                  N       No  
3998                  N      Yes  
3999                  N       No  

[4000 rows x 10 columns]

In [323]:
address_df

address_id  postcode            state    country  property_valuation  \
0              0      2016  New South Wales  Australia                  10   
1              1      2153  New South Wales  Australia                  10   
2              2      4211              QLD  Australia                   9   
3              3      2448  New South Wales  Australia                   4   
4              4      3216              VIC  Australia                   9   
...          ...       ...              ...        ...                 ...   
3995        3995      4511              QLD  Australia                   6   
3996        3996      2756              NSW  Australia                  10   
3997        3997      4032              QLD  Australia                   7   
3998        3998      2251              NSW  Australia                   7   
3999        3999      3996              NSW  Australia                  10   

     house_number               street  
0             060       Morning Avenue  
1               6    Meadow Vale Court  
2               0     Holy Cross Court  
3           17979        Del Mar Point  
4               9       Oakridge Court  
...           ...                  ...  
3995        57042  Village Green Point  
3996           87  Crescent Oaks Alley  
3997         8194          Lien Street  
3998          320          Acker Drive  
3999          613            Erie Lane  

[4000 rows x 7 columns]

In [324]:
product_df

id  product_id           brand product_line product_class product_size  \
0      0           0      OHM Cycles         Road        medium       medium   
1      1           0  Norco Bicycles     Standard           low       medium   
2      2           0      OHM Cycles         Road          high        large   
3      3           0   Trek Bicycles     Standard        medium       medium   
4      4           0      OHM Cycles     Standard        medium       medium   
..   ...         ...             ...          ...           ...          ...   
184  184          98      OHM Cycles     Standard        medium       medium   
185  185          99   Trek Bicycles         Road           low        small   
186  186          99      OHM Cycles     Standard        medium       medium   
187  187         100  Norco Bicycles         Road        medium       medium   
188  188         100   Trek Bicycles     Standard        medium        small   

     list_price  standard_cost  
0        742.54     667.400024  
1        363.01     290.410000  
2         12.01       7.210000  
3        499.53     388.720000  
4        235.63     125.070000  
..          ...            ...  
184      795.34     101.580000  
185     1720.70    1531.420000  
186     1227.34     770.890000  
187     1036.59     206.350000  
188     1386.84    1234.290000  

[189 rows x 8 columns]

### Ненормализованная форма (нулевая нормальная форма)

Исходные таблицы не нарушают базовых принципов реляционной теории - значит можно приступать к нормализации!
(порядок строк и столбцов не имеет значения и не влияет на хранимые данные)

Вывод: все таблицы изначально находятся в нулевой нормальной форме

### Первая нормальная форма

Требования:

* Таблицы находятся в нулевой нормальной форме;
* В таблице не должно быть дубликатов строк;
* В каждой ячейке хранится атомарное значение (не составное);
* В столбцах хранятся данные 1 типа;
* Отсутствуют массивы и списки в любом виде;
* Уникальные названия колонок.

Выводы:

* Все таблицы изначально были в нулевой нормальной форме
* Проверку на дубликаты мы провели на этапе анализа и удалили дубликаты там где они были
* В талице Address мы разделили составное значение адрес на атомарные - дом и улица (в остальных всё ок)
* Во все таблицах в каждом признаке хранятся данные одного типа
* Массивы и списки отсутствуют
* Все таблицы имеют уникальные названия колонок

Итого: все таблицы находятся в 1 нормальной форме!

### Вторая нормальная форма

Требования:

* Все таблицы в 1 нормальной форме
* Все таблицы должны иметь первичный ключ
* Все неключевые столбцы должны неприводимо зависеть от первичного ключа (или от полного - если он составной)

Выводы:

* В каждой таблице есть несоставной первичный ключ (мы их выбрали на этапе анализа, а где было необходимо добавили суррогатные ключи)
* Все неключевые столбцы зависят от первичного ключа так как они у нас везде простые (если потенциальный ключ является простым, то есть состоит из единственного атрибута, то любая функциональная зависимость от него является неприводимой (полной))

Итого: все таблицы находятся во 2 нормальной форме!

### Третья нормальная форма

Требования:

* В таблицах должна отсутствовать транзитивная зависимость

Выводы:

* Так как мы разделили исходные две таблицы transactions и customers на 4 (выделив сущности) мы уже избавились от всех транзитивных зависимостей. Из transactions выделили product, из customers выделили address в отдельные таблицы.

Итого: БД находится в 3 нормальной форме

## Сохраняем таблицы

In [325]:
customer_df.to_csv('customer.csv', index=False)

In [326]:
product_df.to_csv('product.csv', index=False)

In [327]:
address_df.to_csv('address.csv', index=False)

In [328]:
transaction_df.to_csv('transaction.csv', index=False)

## Создаём таблицы в DBeaver

![create_table](https://thumb.cloud.mail.ru/weblink/thumb/xw1/Awfp/13VDrdRoC)

Скрипт сохранил в папке data



## Загрузка данных в DBeaver

При загрузке transaction происходит ошибка из за строки с id 5034, необходимо удалить запись и перезаписать таблицу

In [329]:
transaction_df = transaction_df[transaction_df['customer_id'] != 5034]

In [330]:
transaction_df.to_csv('transaction.csv', index=False)

Transaction

![create_table](https://thumb.cloud.mail.ru/weblink/thumb/xw1/hwck/PjwBBzZuP)

Product

![create_table](https://thumb.cloud.mail.ru/weblink/thumb/xw1/wmsJ/VvmsJtyve)

Customer

![create_table](https://thumb.cloud.mail.ru/weblink/thumb/xw1/6X6S/d5XJvEJmN)

Address

![create_table](https://thumb.cloud.mail.ru/weblink/thumb/xw1/KjgW/iwcJVkeQ7)

DB_diag

![create_table](https://thumb.cloud.mail.ru/weblink/thumb/xw1/FRJp/92JXpDVH3)